### ISSUE TITLE
Problem with custom metric and validation split

### ISSUE THREAD

**Issue by elliottash** [keras-team/keras#10990]
```
from keras import backend as K
def r2(y_true, y_pred):
    SS_res =  K.sum(K.square( y_true-y_pred )) 
    SS_tot = K.sum(K.square( y_true - K.mean(y_true) ) ) 
    return ( 1 - SS_res/(SS_tot + K.epsilon()) )

model.compile(loss='mean_squared_error', # cost function
              optimizer='adam', # use adam as the optimizer
              metrics=[r2_score]) # compute r-squared
model_info = model.fit(X, Yreg, 
                      epochs=3,
                      validation_split=.2)
```

This code produces the following output (check out val_r2):

```
Train on 26053 samples, validate on 6514 samples
Epoch 1/3
26053/26053 [==============================] - 1s 48us/step - loss: 2.3130 - r2: 0.0046 - val_loss: 2.3102 - val_r2: -372309444.3375
Epoch 2/3
26053/26053 [==============================] - 1s 34us/step - loss: 1.9290 - r2: 0.1700 - val_loss: 2.4428 - val_r2: -390677317.6344
Epoch 3/3
26053/26053 [==============================] - 1s 33us/step - loss: 1.8350 - r2: 0.2144 - val_loss: 2.5427 - val_r2: -411575614.3157
```

If I compute r2_score manually (using `sklearn.metrics.r2_score`), the out-of-sample R2 is -3.6.

**Comment by patrickrchao** [comment]
Having the exact same issue, I would like to bump this. I attempted to use the training data as the validation data as a sanity check, and they provide different results.

Model code:

```
def r2(y_true, y_pred):
    SS_res =  K.sum(K.square(y_true - y_pred)) 
    SS_tot = K.sum(K.square(y_true-K.mean(y_true))) 
    return ( 1 - SS_res/(SS_tot + K.epsilon()) )

model = Sequential([
  Dense(64, activation='relu', input_shape=(22,),
  Dense(32, activation='relu',
  Dense(1, activation='linear',
])

model.compile(
  optimizer=optimizers.Adam(lr=1e-2),
  loss="mean_squared_error",
  metrics=[r2]
)

history = model.fit(
  x=X_train,
  y=y_train,
  validation_data=(X_train, y_train),
  epochs=20,
  batch_size=300,
)
```


Training output:
```
Train on 180000 samples, validate on 180000 samples
Epoch 1/20
180000/180000 [==============================] - 2s 9us/step - loss: 0.3979 - r2: 0.0811 - val_loss: 0.3765 - val_r2: -281763.6173
Epoch 2/20
180000/180000 [==============================] - 1s 5us/step - loss: 0.3624 - r2: 0.1626 - val_loss: 0.3465 - val_r2: -182032.9772
Epoch 3/20
180000/180000 [==============================] - 1s 5us/step - loss: 0.3395 - r2: 0.2155 - val_loss: 0.3272 - val_r2: -197065.3954
```

Even with using the training data as the validation data, the validation r^2 is completely off.

**Comment by trent-b** [comment]
I am seeing the same problem. I am using a custom R Squared metric as well. I do not see the problem with a custom Mean Squared Error.

**Comment by Pavlov228** [comment]
The correct implementation of `r2_score` from `sklearn.metrics` is next:

```
def r2(y_true, y_pred):
    SS_res =  K.sum(K.square(y_true - y_pred), axis = 0) 
    SS_tot = K.sum(K.square(y_true - K.mean(y_true, axis = 0)), axis = 0) 
    return K.mean(1 - SS_res/(SS_tot + K.epsilon()), axis = -1 )
```

Note: if you have only 1 output neuron, this function might misbehave. You need to reshape your target vector from `(n_val_examples, )` to `(n_val_examples, 1)` by `numpy.reshape`.

**Comment by rodrigoruiz** [comment]
> The correct implementation of `r2_score` from `sklearn.metrics` is next:
> 
> ```
> def r2(y_true, y_pred):
>     SS_res =  K.sum(K.square(y_true - y_pred), axis = 0) 
>     SS_tot = K.sum(K.square(y_true - K.mean(y_true, axis = 0)), axis = 0) 
>     return K.mean(1 - SS_res/(SS_tot + K.epsilon()), axis = -1 )
> ```
> 
> Note: if you have only 1 output neuron, this function might misbehave. You need to reshape your target vector from `(n_val_examples, )` to `(n_val_examples, 1)` by `numpy.reshape`.

Reshape what and how?
Could you show us the code?

**Comment by rjurney** [comment]
Anyone have something updated for Tensorflow keras? I don't quite get why the sklearn r2_score or the https://www.tensorflow.org/addons/api_docs/python/tfa/metrics/RSquare won't work as loss functions.

**Comment by tavoaqp** [comment]
I'm having the same problem. Weird enough, it only shows the wrong value at training time. When I predict from the hold-out data and run it the R2 score is the same as scikit.
